File: contents/graham_scan/code/java/GrahamScan.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;

public class GrahamScan {

	static class Point {
		public double x;
		public double y;
		
		public Point(double x, double y) {
			this.x = x;
			this.y = y;
		}		
	}
	
	static double ccw(Point a, Point b, Point c) {
		return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
	}
	
	static double polarAngle(Point origin, Point p) {
		return Math.atan2(p.y - origin.y,  p.x - origin.x);
	}
	
	static List<Point> grahamScan(List<Point> gift) {
		gift = new ArrayList<Point>(new HashSet<Point>(gift)); //remove duplicate points
		
		// Sort the list so the point with the lowest y-coordinate comes first
		Comparator<Point> yComparator = (a, b) -> {
			if(b.y - a.y < 0)
				return -1;
			else if(b.y - a.y > 0)
				return 1;
			else return 0;
		};
		
		gift.sort(yComparator);
		Point pivot = gift.get(0);
		
		// Sort the remaining Points based on the angle between the pivot and itself
		List<Point> hull = gift.subList(1, gift.size());
		
		Comparator<Point> angleComparator = (a, b) -> {
			double angleA = polarAngle(a, pivot);
			double angleB = polarAngle(b, pivot);
			if(angleA - angleB < 0)
				return -1;
			else if(angleA - angleB > 0)
				return 1;
			else return 0;
		};
		
		hull.sort(angleComparator);
		
		// The pivot is always on the hull
		hull.add(0, pivot);
		
		int n = hull.size();
		int m = 1;
		for(int i = 2; i < n; i++) {
			while(ccw(hull.get(m - 1), hull.get(m), hull.get(i)) <= 0) {
				if(m > 1) {
					m--;
				} else if(m == 1) {
					break;
				} else {
					i++;
				}
			}
			m++;
			
			Point temp = hull.get(i);
			hull.set(i, hull.get(m));
			hull.set(m, temp);
			
		}
		return hull.subList(0, m + 1);
	}
	

	public static void main(String[] args) {
		ArrayList<Point> points = new ArrayList<Point>();
		
		points.add(new Point(-5, 2));
		points.add(new Point(5, 7));
		points.add(new Point(-6, -12));
		points.add(new Point(-14, -14));
		points.add(new Point(9, 9));
		points.add(new Point(-1, -1));
		points.add(new Point(-10, 11));
		points.add(new Point(-6, 15));
		points.add(new Point(-6, -8));
		points.add(new Point(15, -9));
		points.add(new Point(7, -7));
		points.add(new Point(-2, -9));
		points.add(new Point(6, -5));
		points.add(new Point(0, 14));
		points.add(new Point(2, 8));

		List<Point> convexHull = grahamScan(points);
		
		convexHull.stream().forEach(p -> System.out.printf("% 1.0f, % 1.0f\n", p.x, p.y));
		
	}
	
}

File: contents/graham_scan/code/cpp/graham_scan.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <vector>

struct point {
  double x;
  double y;
};

std::ostream& operator<<(std::ostream& os, const std::vector<point>& points) {
  for (auto p : points) {
    os << "(" << p.x << ", " << p.y << ")\n";
  }
  return os;
}

double ccw(const point& a, const point& b, const point& c) {
  return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}

double polar_angle(const point& origin, const point& p) {
  return std::atan2(p.y - origin.y, p.x - origin.x);
}

std::vector<point> graham_scan(std::vector<point>& points) {
  // selecting lowest point as pivot
  size_t low_index = 0;
  for (size_t i = 1; i < points.size(); i++) {
    if (points[i].y < points[low_index].y) {
      low_index = i;
    }
  }
  std::swap(points[0], points[low_index]);
  point pivot = points[0];

  // sorting points by polar angle
  std::sort(
      points.begin() + 1,
      points.end(),
      [&pivot](const point& pa, const point& pb) {
        return polar_angle(pivot, pa) < polar_angle(pivot, pb);
      });

  // creating convex hull
  size_t m = 1;
  for (size_t i = 2; i < points.size(); i++) {
    while (ccw(points[m - 1], points[m], points[i]) <= 0) {
      if (m > 1) {
        m--;
        continue;
      } else if (i == points.size()) {
        break;
      } else {
        i++;
      }
    }
    m++;
    std::swap(points[i], points[m]);
  }
  return std::vector<point>(points.begin(), points.begin() + m + 1);
}

int main() {
  std::vector<point> points = {{-5, 2},
                               {5, 7},
                               {-6, -12},
                               {-14, -14},
                               {9, 9},
                               {-1, -1},
                               {-10, 11},
                               {-6, 15},
                               {-6, -8},
                               {15, -9},
                               {7, -7},
                               {-2, -9},
                               {6, -5},
                               {0, 14},
                               {2, 8}};
  std::cout << "original points are as follows:\n" << points;
  const std::vector<point> hull = graham_scan(points);
  std::cout << "points in hull are as follows:\n" << hull;
  return 0;
}
